### ISSUE TITLE
Makepot command

### ISSUE THREAD

**Issue by swissspidy** [wp-cli/ideas#68]
Based on the discussion at https://github.com/wp-cli/wp-cli/issues/1042

After https://github.com/wp-cli/wp-cli/pull/1775 added Grunt tasks to `wp plugin scaffold` to generate POT files, the issue was closed.

However, not everyone wants to use Grunt just to use this one task.

> i18n-tools is easy enough to use, so I don't think simply bundling it into WP-CLI would bring much value.

This was a fair point by scribu, but still a Grunt task was added. This means I have the same Grunt task in every plugin, each one bundling i18n-tools again.

There were some concerns afterwards as well, see https://github.com/wp-cli/wp-cli/issues/1042#issuecomment-112980796.

Therefore, I'd like to propose a command like `wp makepot` or `wp language makepot` or similar.

Just like `wp dist-archive` made sure there's only 1 command across the project to create a ZIP file for distribution, there would only be 1 global command to create a POT file.

If there's enough interest in this I'd help kick things off with a proof-of-concept command.

**Comment by schlessera** [comment]
I think one important advantage of this would be that, contrary to the grunt solution, you could create POT files in an external process, where the tools don't need to be part of the plugin already. This way, you could even update bad or missing POT files for existing plugins as well.

**Comment by danielbachhuber** [comment]
I'm 👍 to this idea.

**Comment by danielbachhuber** [comment]
Would this be bundled with WP-CLI or offered standalone?

**Comment by swissspidy** [comment]
Well, if it's possible to bundle it with WP-CLI I wouldn't say no to that :-) I think it could be very useful for many developers.

I started work on it here: https://github.com/swissspidy/wp-cli-makepot

Unfortunately the i18n tools only have a few tests and badly formatted code. Maybe this is an opportunity to add some great improvements there. (@ocean90?)

**Comment by bradyvercher** [comment]
I stumbled across this issue and, as the developer of the Grunt task, I'm all for a WP-CLI command. I also figured I'd provide a little insight into how the Grunt task works and clear up some misconceptions.

Regarding the concerns in wp-cli/wp-cli#1042, Grunt isn't used to interpret PHP. It's actually an interface to the same i18n tools mentioned here. The task just scans the project to determine some sensible defaults and provides a way to configure the task, but then passes the arguments on to the PHP tools. Unfortunately, the PHP tools are a bit limited, so I had to fork them early on to add some extra options.

Earlier this year, I extracted the core of the Grunt task into [node-wp-i18n](https://github.com/cedaro/node-wp-i18n) and made Grunt a wrapper for that project. Grunt is completely optional. The node package also provides [a CLI interface](https://github.com/cedaro/node-wp-i18n/blob/develop/docs/cli.txt), so it can be installed globally and doesn't need to be included in every project.

I've considered extracting the forked tools back out into a PHP-only repo and writing a WP-CLI command in the past, but the existing packages have done everything I need.

I think a canonical source for the PHP tools that's easy to install or bundle would be beneficial so that improvements could be easily propagated to everyone no matter which task runner they wanted to use. If this issue produces a PHP package that incorporates the improvements from the forked version in `node-wp-i18n`, I'd be all for bundling that instead.

**Comment by swissspidy** [comment]
Thanks for your insights!

I also talked with ocean90 in the meantime about creating a single PHP package that can be used by everyone. He pointed out that the current i18n-tools are heavily focused on WordPress.org and their infrastructure. Since this WP-CLI task would only be needed for plugins and themes, we don't need lots of things like creating POT files for WP Core or Rosetta sites, `guess_plugin_slug()`, `get_addon_header()` (could use the WP functions), etc.
tl;dr: we should roll our own instead of trying to make a canonical i18n-tools repo out of it.

According to him, the only thing that's truly unique and that could be put in its own repository is the extraction part (https://github.com/ocean90/wp-i18n-tools/blob/master/extract/extract.php). However, I don't think we need to reinvent the wheel there. There are powerful libraries like https://packagist.org/packages/gettext/gettext that can be used for string extraction.

**Comment by bradyvercher** [comment]
No problem, @swissspidy! I didn't meant to suggest the new package should be a copy of the i18n tools. Stripping out some of the cruft and streamlining it would definitely make sense.

I guess the approach really depends on if WP-CLI is going to be the only interface to the library. If it is, then WP functions could be used, but for something more broadly useful, those functions won't be available. For instance, if I wanted to declare the package as a dev dependency in a plugin's `composer.json` and add a script to build the POT file, it shouldn't rely on a WordPress environment.

I think using a third-party library like [gettext/gettext](https://packagist.org/packages/gettext/gettext) should be fine for extracting strings, but there are a few things that are unique to WordPress that would need to be handled separately. Plugin and theme file headers, page template headers, and translator comments come to mind.

The original PHP i18n tools also depend on having gettext installed to use utilities like [msguniq](https://github.com/ocean90/wp-i18n-tools/blob/master/makepot.php#L502) for removing duplicate strings. The node package I wrote [works around that requirement](https://github.com/cedaro/grunt-wp-i18n/issues/2), so ideally something should be possible in PHP, too.

**Comment by swissspidy** [comment]
Hi all,

I've been digging into https://github.com/swissspidy/wp-cli-makepot again today and decided to give that [gettext/gettext](https://packagist.org/packages/gettext/gettext) library a shot and kick out the WordPress i18n-tools.

The library supports things like translator comments out of the box, so I basically just needed to add some code to handle file headers and page templates.

After removing all the clutter, I've reduced the command to `wp makepot plugin` and `wp makepot theme`, as these are the only things a WP developer really needs.

I still have to write some integration tests and make some tweaks here and there, but it seems to work fine so far already.

**Comment by swissspidy** [comment]
This is being worked on in https://github.com/wp-cli/i18n-command
